pl-faded-parsons.js

```diff
@@ -1,377 +1,713 @@
+/** expects a config with a `uuid` and fields that align with this schema:
+ * ```
+ *  ...
+ *  <div#{{config.main}}.fpp-global-defs>
+ *    <!-- inputs in the template that will save
+ *         the student's progress between reloads
+ *    -->
+ *    <input#{{config.solutionOrderStorage}}
+ *    <input#{{config.starterOrderStorage}}
+ *   ...
+ *   ( <!-- A starter tray is optional -->
+ *      <div#{{config.starter}}.codeline-tray>
+ *         <ul#{{config.starterList}}.codeline-list>
+ *           <li.codeline> ... <input.parsons-blank>* ... </li>*
+ *         </ul>
+ *       </div>
+ *   )?
+ *    ...
+ *    <div#{{config.solution}}.codeline-tray>
+ *       <ul#{{config.solutionList}}.codeline-list>
+ *         <li.codeline> ... <input.parsons-blank>* ... </li>*
+ *       </ul>
+ *    </div>
+ *     ...
+ *    <div#{{config.toolbar}>
+ *       ...
+ *       <div.fpp-help data-toggle="popover"></div>
+ *       <div.fpp-copy></div>
+ *       <div.fpp-dark></div>
+ *       ...
+ *    </div>
+ *     ...
+ *  </div>
+ * ```
+ */
+class ParsonsWidget {
+  /** Creates a new widget instance. See class docs for info on `config`. */
+  constructor(config) {
+    this.config = jQuery.extend(
+      {
+        xIndent: 4,
+        canIndent: true,
+        prettyPrint: true,
+        onSortableUpdate: (_event, _ui) => {},
+        onBlankUpdate: (_event, _input) => {},
+        lang: "en",
+      },
+      config,
+    );
+
+    /** When true, navigating to a codeline with arrow keys enters its first blank */
+    this.autoEnterBlank = true;
+
+    const requiredFields =
+      "uuid solution solutionList main toolbar starterOrderStorage solutionOrderStorage solutionSubmissionStorage";
+    const missing = requiredFields
+      .split(" ")
```

Comment:
Potential nitpick: Does `requiredFields` have to be a string? Why isn't it an array literal?